File: Data Structures & Algorithms/balanced-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    getHeight(node) {
        if(node === null) return 0
        return 1 + Math.max(this.getHeight(node.left), this.getHeight(node.right))
    }

    isBalanced(root) {
        if(root === null) return true

        if(Math.abs(this.getHeight(root.left) - this.getHeight(root.right)) > 1) {
            return false
        }
        
        return this.isBalanced(root.left) && this.isBalanced(root.right)        
    }
}

File: Data Structures & Algorithms/insert-into-a-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if not root:
            return TreeNode(val=val)
        r = root
        while root:
            if val < root.val and root.left:
                root = root.left
            elif val > root.val and root.right:
                root = root.right
            elif val < root.val and not root.left:
                root.left = TreeNode(val=val)
                break
            elif val > root.val and not root.right:
                root.right = TreeNode(val=val)
                break
        return r